tests/integration/security/reachability/context.go

```diff
@@ -0,0 +1,171 @@
+
+	// Indicates whether a test should be created for the given configuration.
+	Include func(src echo.Instance, opts echo.CallOptions) bool
+
+	// Handler called when the given test is being run.
+	OnRun func(ctx framework.TestContext, src echo.Instance, opts echo.CallOptions)
+
+	// Indicates whether the test should expect a successful response.
+	ExpectSuccess func(src echo.Instance, opts echo.CallOptions) bool
+}
+
+// Context is a context for reachability tests.
+type Context struct {
+	ctx                   framework.TestContext
+	g                     galley.Instance
+	p                     pilot.Instance
+	Namespace             namespace.Instance
+	A, B, Headless, Naked echo.Instance
+}
+
+// CreateReachabilityContext creates and initializes reachability context.
+func CreateContext(ctx framework.TestContext, g galley.Instance, p pilot.Instance) Context {
+	ns := namespace.NewOrFail(ctx, ctx, namespace.Config{
+		Prefix: "reachability",
+		Inject: true,
+	})
+
+	var a, b, headless, naked echo.Instance
+	echoboot.NewBuilderOrFail(ctx, ctx).
+		With(&a, util.EchoConfig("a", ns, false, nil, g, p)).
+		With(&b, util.EchoConfig("b", ns, false, nil, g, p)).
+		With(&headless, util.EchoConfig("headless", ns, false, nil, g, p)).
+		With(&naked, util.EchoConfig("naked", ns, false, echo.NewAnnotations().
+			SetBool(echo.SidecarInject, false), g, p)).
+		BuildOrFail(ctx)
+
+	return Context{
+		ctx:       ctx,
+		g:         g,
+		p:         p,
+		Namespace: ns,
+		A:         a,
+		B:         b,
+		Headless:  headless,
+		Naked:     naked,
+	}
+}
+
+// Run runs the given reachability test cases with the context.
+func (rc *Context) Run(testCases []TestCase) {
+	callOptions := []echo.CallOptions{
+		{
+			PortName: "http",
+			Scheme:   scheme.HTTP,
+		},
+		{
+			PortName: "http",
+			Scheme:   scheme.WebSocket,
+		},
+		{
+			PortName: "tcp",
+			Scheme:   scheme.HTTP,
+		},
+		{
+			PortName: "grpc",
+			Scheme:   scheme.GRPC,
+		},
+	}
+
+	for _, c := range testCases {
+		c := c
+		testName := strings.TrimSuffix(c.ConfigFile, filepath.Ext(c.ConfigFile))
+		test := rc.ctx.NewSubTest(testName)
+
+		if c.RequiredEnvironment != "" {
+			test.RequiresEnvironment(c.RequiredEnvironment)
+		}
+
+		test.Run(func(ctx framework.TestContext) {
+			// Apply the policy.
+			policyYAML := file.AsStringOrFail(ctx, filepath.Join("../testdata", c.ConfigFile))
+			rc.g.ApplyConfigOrFail(rc.ctx, c.Namespace, policyYAML)
+			rc.ctx.WhenDone(func() error {
+				return rc.g.DeleteConfig(c.Namespace, policyYAML)
+			})
+
+			// Give some time for the policy propagate.
+			// TODO: query pilot or app to know instead of sleep.
+			time.Sleep(10 * time.Second)
+
+			for _, src := range []echo.Instance{rc.A, rc.B, rc.Headless, rc.Naked} {
+				for _, dest := range []echo.Instance{rc.A, rc.B, rc.Headless, rc.Naked} {
+					for _, opts := range callOptions {
+						// Copy the loop variables so they won't change for the subtests.
+						src := src
```

Comment:
I prefer we either rename the var or do param binding. same var name is kinda of hard to follow

```go
func(a string, b string) {
}(src, dest)
```

Reply:
This is the same code as before, so I'd like not to change to avoid any side effect.